packages/portfolio-deploy/src/vstorage-prune.build.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * @file enumerate [vstorage][vsmod] keys under `published.ymax` that are
+ * older than the ymax0 instance in published.agoricNames.instance
+ *
+ * Motivation: each time we deploy to devnet, portfolios and such from
+ * earlier deployments remain in vstorage, confusing clients.
+ *
+ * [vsmod]: https://github.com/Agoric/agoric-sdk/blob/ymax-v0.1-alpha/golang/cosmos/x/vstorage/README.md
+ */
+/* global globalThis */
+import { fetchEnvNetworkConfig, makeVStorage } from '@agoric/client-utils';
+import { makeHelpers } from '@agoric/deploy-script-support';
+import { NonNullish } from '@agoric/internal';
+import { Fail } from '@endo/errors';
+import { getManifestForPruneVStorage } from './vstorage-prune.core.js';
+
+const sourceSpec = './vstorage-prune.core.js';
+
+/**
+ * @import { StreamCell } from '@agoric/internal/src/lib-chainStorage.js';
+ * @import { VStorage } from '@agoric/client-utils';
+ * @import {CoreEvalBuilder, DeployScriptFunction} from '@agoric/deploy-script-support/src/externalTypes.js';
+ * @import {PruneOpts} from './vstorage-prune.core';
+ */
+
+/**
+ * @param {unknown} _utils
+ * @param {PruneOpts} pruneOpts
+ * @satisfies {CoreEvalBuilder}
+ */
+export const defaultProposalBuilder = async (_utils, pruneOpts) =>
+  harden({
+    sourceSpec,
+    getManifestCall: [getManifestForPruneVStorage.name, { options: pruneOpts }],
+  });
+
+/**
+ * @param {VStorage} vs
+ * @param {string} start
+ */
+async function* depthFirst(vs, start) {
+  /** @param {string} current */
+  async function* visit(current) {
+    const children = await vs.keys(current);
+    for (const child of children) {
+      yield* visit(`${current}.${child}`);
+    }
+    yield current;
+  }
+
+  yield* visit(start);
+}
+
+/** @param {VStorage} vs */
+const findOutdated = async vs => {
+  const verbose = false; // XXX cli flag
+  const debug = verbose ? console.debug : () => {};
+
+  /** @param {string} path */
+  const getHeightLast = async path => {
+    debug('read', path);
+    // const raw = await vs.readStorage(path);
+    // console.debug({ path, raw });
+
+    const { blockHeight, values } = /** @type {StreamCell<'string'>} */ (
+      await vs.readAt(path)
+    );
+    return { blockHeight: Number(blockHeight), last: values.at(-1) };
+  };
+
+  const { blockHeight: t0, last } = await getHeightLast(
+    'published.agoricNames.instance',
+  );
+  (last && last.includes('ymax0')) || Fail`no ymax0 instance`;
+  console.error('published.agoricNames.instance blockHeight', t0);
+
+  /** @type {PruneOpts['toPrune']} */
+  const toPrune = {};
+
+  const noData = ['.portfolios', '.flows', '.positions'];
```

Comment:
Optional: Are these keys able to be imported from somewhere? Also consider putting 'published.ymax0.portfolios' behind a descriptively named const

Reply:
I suppose `type-guards.ts` should export this info. I think I already have an XXX there about its API being awkward. I hope to clean it up later.